src/flow/src/compute/types.rs

```diff
@@ -0,0 +1,143 @@
+    ///
+    /// Note: need to be the same hydroflow instance that this collection is created from
+    pub fn clone(&self, df: &mut Hydroflow) -> Self {
+        Collection {
+            stream: self.stream.tee(df),
+        }
+    }
+}
+
+/// Arranged is a wrapper around `ArrangeHandler` that maintain a list of readers and a writer
+pub struct Arranged {
+    pub arrangement: ArrangeHandler,
+    pub writer: Rc<RefCell<Option<SubgraphId>>>,
+    /// maintain a list of readers for the arrangement for the ease of scheduling
+    pub readers: Rc<RefCell<Vec<SubgraphId>>>,
+}
+
+impl Arranged {
+    pub fn new(arr: ArrangeHandler) -> Self {
+        Self {
+            arrangement: arr,
+            writer: Default::default(),
+            readers: Default::default(),
+        }
+    }
+    pub fn try_clone_future(&self) -> Option<Self> {
+        self.arrangement
+            .clone_future_only()
+            .map(|arrangement| Arranged {
+                arrangement,
+                readers: self.readers.clone(),
+                writer: self.writer.clone(),
+            })
+    }
+    pub fn try_clone_full(&self) -> Option<Self> {
+        self.arrangement
+            .clone_full_arrange()
+            .map(|arrangement| Arranged {
+                arrangement,
+                readers: self.readers.clone(),
+                writer: self.writer.clone(),
+            })
+    }
+    pub fn add_reader(&self, id: SubgraphId) {
+        self.readers.borrow_mut().push(id)
+    }
+}
+
+/// A bundle of the various ways a collection can be represented.
+///
+/// This type maintains the invariant that it does contain at least one(or both) valid
+/// source of data, either a collection or at least one arrangement. This is for convenience
+/// of reading the data from the collection.
+pub struct CollectionBundle {
+    pub collection: Collection<DiffRow>,
+    /// the key [`ScalarExpr`] indicate how the keys(also a [`Row`]) used in Arranged is extract from collection's [`Row`]
+    /// So it is the "index" of the arrangement
+    pub arranged: BTreeMap<Vec<ScalarExpr>, Arranged>,
+}
+
+impl CollectionBundle {
+    pub fn from_collection(collection: Collection<DiffRow>) -> Self {
+        Self {
+            collection,
+            arranged: BTreeMap::default(),
+        }
+    }
+    pub fn clone(&self, df: &mut Hydroflow) -> Self {
+        Self {
+            collection: self.collection.clone(df),
+            arranged: self
+                .arranged
+                .iter()
+                .map(|(k, v)| (k.clone(), v.try_clone_future().unwrap()))
+                .collect(),
+        }
+    }
+}
+
+/// A thread local error collector, used to collect errors during the evaluation of the plan
+///
+/// usually only the first error matters, but store all of them just in case
+#[derive(Default, Clone)]
+pub struct ErrCollector {
+    pub inner: Rc<RefCell<VecDeque<EvalError>>>,
```

Comment:
Why `VecDeque` here. How will those stored errors be used later

Reply:
> Why `VecDeque` here. How will those stored errors be used later

The original idea was to use error collector to continous collect those non-lethal `EvalError` and output them to somewhere(i.e. log) while perserve their order